src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -10,6 +10,14 @@ namespace System.Net.Sockets
 {
     public partial class Socket
     {
+        // TODO: make internal
+        public static Socket CreateForPipeSafeHandle(IntPtr handle, bool ownsHandle)
```

Comment:
@stephentoub do you have a suggestion how I can make this `internal` but still usable from `System.IO.Pipes` assembly?

Reply:
This is needed to enable only PipeStream to use the SafeSocketHandle(int fd, ...) ctor with a pipe but no one else?

With these changes I think we could remove some of those restrictions.  If the SafeSocketHandle can then actually be used with non-sockets, at least for some operations like reads and writes, I think it'd be fine to just let the existing SafeSocketHandle be constructed with such descriptors and removing the logic that throws if it's not a socket.

What would break if we did that?

Reply:
Nothing would break.

The question is whether it is something we want end-users to be capable of doing using the public API?

Reply:
> The question is whether it is something we want end-users to be capable of doing using the public API?

If nothing would break, I don't think we need to go out of our way to prevent it.  We added the "is this a socket" check out of an abundance of caution, but if everything will "just work" (or if socket operations that can't work will just fail with a reasonable error), I don't see a strong need to cut off valid uses.  FileStream, for example, on both Windows and Linux can be created with SafeFileHandles that aren't disk files.

@geoffkizer, @wfurt, any concerns with loosening the restrictions here?

Reply:
I think it would be good IMHO to allow Socket-ish interface on other types where OS supports it. Aside from pipes, some entries in /dev, /proc and /sys may benefit from it on Linux. It may be hard for us to cover and test all the weird cases but at least for pipes I think we can make get decent test coverage.

Reply:
I've changed it to using the public constructor.